### ISSUE TITLE
`@avx` is extremely slower than manually expanded loop

### ISSUE THREAD

**Issue by Roger-luo** [JuliaSIMD/LoopVectorization.jl#189]
It's a bit to my surprise that in the following vector operation `@avx` is about 10x slower than a manual loop unroll

```julia
using BenchmarkTools
using LoopVectorization

function broutine_z!(st::Vector{Complex{T}}, ::Val{:Z}, locs) where T
    step_1 = 1 << (locs - 1)
    step_2 = 1 << locs

    @inbounds for j in 0:step_2:size(st, 1)-step_1
        for i in j:j+step_1-1
            ST1 = st[i+1]
            ST2 = -st[i+step_1+1]
        
            st[i+1] = ST1
            st[i+step_1+1] = ST2
        end
    end
    return st
end

function broutine_z_expanded!(st::Vector{Complex{T}}, ::Val{:Z}, locs) where T
    step_1 = 1 << (locs - 1)
    step_2 = 1 << locs

    @inbounds for j in 0:step_2:size(st, 1)-step_1
        for i in j:16:j+step_1-1
            Base.Cartesian.@nexprs 16 k->begin
                ST1 = st[i+k]
                ST2 = -st[i+step_1+k]
            
                st[i+k] = ST1
                st[i+step_1+k] = ST2
            end
        end
    end
    return st
end

function broutine_z_avx!(st::Vector{Complex{T}}, ::Val{:Z}, locs) where T
    mshift = locs
    mstride = 1 << mshift
    nstride = 1 << (mshift - 1)
    Mmax = size(st, 1)
    str = reinterpret(reshape, T, st)
    M = (Mmax-1) >>> mshift
    @avx for _m in 0:M
        m = _m << mshift
        for n in 1:nstride
            T1_re = str[1, m+n]
            T1_im = str[2, m+n]
            T2_re = -str[1, m+n+nstride]
            T2_im = -str[2, m+n+nstride]

            str[1, m+n] = T1_re
            str[2, m+n] = T1_im
            str[1, m+n+nstride] = T2_re
            str[2, m+n+nstride] = T2_im
        end
    end
    return st
end

st=rand(ComplexF64, 1 << 20);
broutine_z_avx!(copy(st), Val(:Z), 19) ≈ broutine_z!(copy(st), Val(:Z), 19)
broutine_z_expanded!(copy(st), Val(:Z), 19) ≈ broutine_z!(copy(st), Val(:Z), 19)

@benchmark broutine_z!(st, Val(:Z), 19) setup=(st=rand(ComplexF64, 1 << $n))
@benchmark broutine_z_avx!(st, Val(:Z), 19) setup=(st=rand(ComplexF64, 1 << $n))
@benchmark broutine_z_expanded!(st, Val(:Z), 19) setup=(st=rand(ComplexF64, 1 << $n))
```

gives

```julia
julia> @benchmark broutine_z!(st, Val(:Z), 19) setup=(st=rand(ComplexF64, 1 << $n))
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     289.085 μs (0.00% GC)
  median time:      333.229 μs (0.00% GC)
  mean time:        339.611 μs (0.00% GC)
  maximum time:     421.622 μs (0.00% GC)
  --------------
  samples:          1400
  evals/sample:     1

julia> @benchmark broutine_z_avx!(st, Val(:Z), 19) setup=(st=rand(ComplexF64, 1 << $n))
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     2.078 ms (0.00% GC)
  median time:      2.102 ms (0.00% GC)
  mean time:        2.103 ms (0.00% GC)
  maximum time:     2.383 ms (0.00% GC)
  --------------
  samples:          943
  evals/sample:     1

julia> @benchmark broutine_z_expanded!(st, Val(:Z), 19) setup=(st=rand(ComplexF64, 1 << $n))
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     131.527 μs (0.00% GC)
  median time:      196.361 μs (0.00% GC)
  mean time:        204.807 μs (0.00% GC)
  maximum time:     307.005 μs (0.00% GC)
  --------------
  samples:          1444
  evals/sample:     1
```

I'm running Julia-1.6, I think based on #169 I have reinterpreted the array to `HWReal` type, so `@avx` should work here I assume? Replacing `@avx` with just `@inbounds` seems to give a not bad performance.

---

It seems somehow `@avx` slows down the loop.

**Comment by chriselrod** [comment]
This is caused by a problem I'll address in the rewrite.
The problem was discussed [here](https://github.com/chriselrod/LoopVectorization.jl/issues/132).

Basically, at the moment, LoopVectorization just uses gathers & scatters to access non-contiguous elements. E.g., on a system with AVX512:
```asm
# julia> @code_native debuginfo=:none syntax=:intel broutine_z_avx!(st, Val(:Z), 19)
        .text
        push    rbp
        push    r15
        push    r14
        push    r13
        push    r12
        push    rbx
        mov     rax, rdi
        lea     rdx, [rsi - 1]
        mov     edi, 1
        shlx    rbp, rdi, rdx
        xor     ecx, ecx
        cmp     rdx, 63
        cmova   rbp, rcx
        mov     ebx, 1
        sub     rbx, rsi
        shrx    r10, rdi, rbx
        cmp     rbx, 63
        cmova   r10, rcx
        test    rdx, rdx
        cmovns  r10, rbp
        mov     rdx, qword ptr [rax + 24]
        dec     rdx
        shrx    rdi, rdx, rsi
        cmp     rsi, 63
        cmova   rdi, rcx
        mov     r11, qword ptr [rax]
        mov     rbx, rsi
        neg     rbx
        shlx    r8, rdx, rbx
        cmp     rbx, 63
        cmova   r8, rcx
        mov     r9d, 63
        cmovb   r9, rbx
        test    rsi, rsi
        cmovns  r8, rdi
        lea     r15, [r11 + 8]
        mov     ecx, r10d
        neg     cl
        and     cl, 7
        mov     dl, -1
        shr     dl, cl
        kmovd   k1, edx
        cmp     r10, 8
        jge     L402
        inc     r8
        xor     ecx, ecx
        movabs  rdx, offset .rodata.cst32
        vmovapd ymm0, ymmword ptr [rdx]
        movabs  rdx, offset .rodata.cst8
        vbroadcastsd    zmm1, qword ptr [rdx]
        nop     dword ptr [rax]
L192:
        cmp     rsi, 63
        shlx    rdx, rcx, rsi
        mov     edi, 0
        cmovbe  rdi, rdx
        test    rsi, rsi
        sarx    rdx, rcx, r9
        cmovns  rdx, rdi
        mov     rdi, rdx
        shl     rdi, 4
        lea     rbx, [r11 + rdi]
        vxorpd  xmm2, xmm2, xmm2
        kmovq   k2, k1
        vgatherdpd      zmm2 {k2}, zmmword ptr [rbx + 8*ymm0]
        add     rdi, r15
        vxorpd  xmm3, xmm3, xmm3
        kmovq   k2, k1
        vgatherdpd      zmm3 {k2}, zmmword ptr [rdi + 8*ymm0]
        add     rdx, r10
        shl     rdx, 4
        lea     rbp, [r11 + rdx]
        vxorpd  xmm4, xmm4, xmm4
        kmovq   k2, k1
        vgatherdpd      zmm4 {k2}, zmmword ptr [rbp + 8*ymm0]
        add     rdx, r15
        vxorpd  xmm5, xmm5, xmm5
        kmovq   k2, k1
        vgatherdpd      zmm5 {k2}, zmmword ptr [rdx + 8*ymm0]
        kmovq   k2, k1
        vscatterdpd     zmmword ptr [rbx + 8*ymm0] {k2}, zmm2
        kmovq   k2, k1
        vscatterdpd     zmmword ptr [rdi + 8*ymm0] {k2}, zmm3
        vxorpd  zmm2, zmm4, zmm1
        kmovq   k2, k1
        vscatterdpd     zmmword ptr [rbp + 8*ymm0] {k2}, zmm2
        vxorpd  zmm2, zmm5, zmm1
        kmovq   k2, k1
        vscatterdpd     zmmword ptr [rdx + 8*ymm0] {k2}, zmm2
        inc     rcx
        cmp     r8, rcx
        jne     L192
L388:
        pop     rbx
        pop     r12
        pop     r13
        pop     r14
        pop     r15
        pop     rbp
        vzeroupper
        ret
L402:
        lea     r13, [r10 - 8]
        mov     r12, -1
        xor     r14d, r14d
        movabs  rcx, offset .rodata.cst32
        vmovapd ymm0, ymmword ptr [rcx]
        movabs  rcx, offset .rodata.cst8
        vbroadcastsd    zmm1, qword ptr [rcx]
        jmp     L453
L448:
        cmp     r12, r8
        je      L388
L453:
        inc     r12
        cmp     rsi, 63
        shlx    rcx, r12, rsi
        cmova   rcx, r14
        test    rsi, rsi
        sarx    rdx, r12, r9
        cmovns  rdx, rcx
        lea     rbx, [r10 + rdx]
        shl     rbx, 4
        lea     rdi, [r15 + rbx]
        add     rbx, r11
        shl     rdx, 4
        lea     rcx, [r15 + rdx]
        add     rdx, r11
        xor     ebp, ebp
        nop     dword ptr [rax]
L512:
        kxnorw  k2, k0, k0
        vgatherdpd      zmm2 {k2}, zmmword ptr [rdx + 8*ymm0]
        kxnorw  k2, k0, k0
        vgatherdpd      zmm3 {k2}, zmmword ptr [rcx + 8*ymm0]
        kxnorw  k2, k0, k0
        vgatherdpd      zmm4 {k2}, zmmword ptr [rbx + 8*ymm0]
        kxnorw  k2, k0, k0
        vgatherdpd      zmm5 {k2}, zmmword ptr [rdi + 8*ymm0]
        kxnorw  k2, k0, k0
        vscatterdpd     zmmword ptr [rdx + 8*ymm0] {k2}, zmm2
        kxnorw  k2, k0, k0
        vscatterdpd     zmmword ptr [rcx + 8*ymm0] {k2}, zmm3
        vxorpd  zmm2, zmm4, zmm1
        kxnorw  k2, k0, k0
        vscatterdpd     zmmword ptr [rbx + 8*ymm0] {k2}, zmm2
        vxorpd  zmm2, zmm5, zmm1
        kxnorw  k2, k0, k0
        vscatterdpd     zmmword ptr [rdi + 8*ymm0] {k2}, zmm2
        add     rbp, 8
        sub     rdi, -128
        sub     rbx, -128
        sub     rcx, -128
        sub     rdx, -128
        cmp     rbp, r13
        jle     L512
        cmp     rbp, r10
        jge     L448
        kmovq   k2, k1
        vxorpd  xmm2, xmm2, xmm2
        vgatherdpd      zmm2 {k2}, zmmword ptr [rdx + 8*ymm0]
        kmovq   k2, k1
        vxorpd  xmm3, xmm3, xmm3
        vgatherdpd      zmm3 {k2}, zmmword ptr [rcx + 8*ymm0]
        kmovq   k2, k1
        vxorpd  xmm4, xmm4, xmm4
        vgatherdpd      zmm4 {k2}, zmmword ptr [rbx + 8*ymm0]
        kmovq   k2, k1
        vxorpd  xmm5, xmm5, xmm5
        vgatherdpd      zmm5 {k2}, zmmword ptr [rdi + 8*ymm0]
        kmovq   k2, k1
        vscatterdpd     zmmword ptr [rdx + 8*ymm0] {k2}, zmm2
        kmovq   k2, k1
        vscatterdpd     zmmword ptr [rcx + 8*ymm0] {k2}, zmm3
        vxorpd  zmm2, zmm4, zmm1
        kmovq   k2, k1
        vscatterdpd     zmmword ptr [rbx + 8*ymm0] {k2}, zmm2
        vxorpd  zmm2, zmm5, zmm1
        kmovq   k2, k1
        vscatterdpd     zmmword ptr [rdi + 8*ymm0] {k2}, zmm2
        jmp     L448
        nop     word ptr [rax + rax]
```
These are very slow. What LLVM does (and what LoopVectorization will do) is replace these gathers/scatters with contiguous loads/stores mixed with permutes/shuffles, when it is possible to do so.
That is much faster, but still not as fast as expanded.

Because you aren't treating real and imaginary numbers differently, it's even faster to just skip the permutes/shuffling.

There was also a secondary bug in LoopVectorization, where it wasn't check for `A[m << 1]` type loads not being contiguous if you SIMD `m`. This is fixed in LoopVectorization 0.9.16, which I'll tag for a release after the tests pass.

Lets mimic that
```julia
function broutine_z_flat!(st::Vector{Complex{T}}, ::Val{:Z}, locs) where T
    str = reinterpret(T, st) # no reshape
    step_1 = 1 << locs
    step_2 = 1 << (locs + 1)

    @inbounds for j in 0:step_2:size(str, 1)-step_1
        for i in j:j+step_1-1
            ST1 = str[i+1]
            ST2 = -str[i+step_1+1]
        
            str[i+1] = ST1
            str[i+step_1+1] = ST2
        end
    end
    return st
end
function broutine_z_flat_avx!(st::Vector{Complex{T}}, ::Val{:Z}, locs) where T
    mshift = locs
    nstride = 1 << mshift
    str = reinterpret(T, st) # no reshape
    Mmax = size(st, 1)
    M = (Mmax-1) >>> mshift
    @avx for _m in 0:M
        m = _m << (mshift + 1)
        for n in 1:nstride
            T1 = str[m+n]
            T2 = -str[m+n+nstride]

            str[m+n] = T1
            str[m+n+nstride] = T2
        end
    end
    return st
end
```
Also, I assume this is a simplified problem, because 
```julia
function broutine_z!(st::Vector{Complex{T}}, ::Val{:Z}, locs) where T
    step_1 = 1 << (locs - 1)
    step_2 = 1 << locs

    @inbounds for j in 0:step_2:size(st, 1)-step_1
        for i in j:j+step_1-1
            ST1 = st[i+1]
            ST2 = -st[i+step_1+1]
        
            st[i+1] = ST1
            st[i+step_1+1] = ST2
        end
    end
    return st
end
```
Can be simplified to:
```julia
function broutine_z_simple!(st::Vector{Complex{T}}, ::Val{:Z}, locs) where T
    step_1 = 1 << (locs - 1)
    step_2 = 1 << locs

    @inbounds for j in 0:step_2:size(st, 1)-step_1
        for i in j:j+step_1-1
            st[i+step_1+1] *= -1
        end
    end
    return st
end
```
Confirming:
```julia
julia> st2 = copy(st);

julia> broutine_z_simple!(st2, Val(:Z), 19) == broutine_z!(st, Val(:Z), 19)
true
```
(LLVM unnecessarily uses permutes there because it doesn't do the flattening of complex numbers on its own, so this version is not the fastest.)

**Comment by chriselrod** [comment]
Should perform about the same on LoopVectorization 0.12 now:
```julia
julia> st=rand(ComplexF64, 1 << 20);

julia> broutine_z_avx!(copy(st), Val(:Z), 19) ≈ broutine_z!(copy(st), Val(:Z), 19)
true

julia> broutine_z_expanded!(copy(st), Val(:Z), 19) ≈ broutine_z!(copy(st), Val(:Z), 19)
true

julia> n = 19;

julia> @benchmark broutine_z!(st, Val(:Z), 19) setup=(st=rand(ComplexF64, 1 << $n))
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     181.597 μs (0.00% GC)
  median time:      305.238 μs (0.00% GC)
  mean time:        304.483 μs (0.00% GC)
  maximum time:     391.242 μs (0.00% GC)
  --------------
  samples:          2513
  evals/sample:     1

julia> @benchmark broutine_z_avx!(st, Val(:Z), 19) setup=(st=rand(ComplexF64, 1 << $n))
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     183.313 μs (0.00% GC)
  median time:      306.207 μs (0.00% GC)
  mean time:        303.356 μs (0.00% GC)
  maximum time:     351.649 μs (0.00% GC)
  --------------
  samples:          2509
  evals/sample:     1

julia> @benchmark broutine_z_expanded!(st, Val(:Z), 19) setup=(st=rand(ComplexF64, 1 << $n))
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     170.602 μs (0.00% GC)
  median time:      298.659 μs (0.00% GC)
  mean time:        286.407 μs (0.00% GC)
  maximum time:     362.063 μs (0.00% GC)
  --------------
  samples:          2378
  evals/sample:     1
```
